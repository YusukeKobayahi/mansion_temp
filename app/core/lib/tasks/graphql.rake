# frozen_string_literal: true

module GraphqlTask
  extend Rake::DSL

  namespace :graphql do
    namespace :consumer do
      task schema_dump: :environment do
        require 'graphql/rake_task'

        GraphQL::RakeTask.new(
          load_schema: lambda { |_task|
            require File.expand_path('../../app/graphql/consumer/inno_schema', __dir__)
            Consumer::InnoSchema
          },
          directory: './graphql/consumer',
        )
        Rake::Task['graphql:schema:idl'].invoke
      end
    end
  end
end
