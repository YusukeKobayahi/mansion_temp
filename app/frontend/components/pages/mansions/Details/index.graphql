fragment MsLibraryMansionsDetails on Mansion {
  name
  age
  constructedIn
  blockNumber
  groundFloor
  undergroundFloor
  unitAmount
  landPrivilege
  constructCompany
  managementForm
  managementCompany
  structure {
    name
  }
  prefecture {
    name
    jisCode
  }
  city {
    name
    jisCode
  }
  town {
    id
    name
  }
  street {
    name
  }
  access {
    id
    stationId
    stationLineName
    stationName
    stationWalkingMinutes
    lineId
  }
  parking
  landUseZones
  layouts
}
