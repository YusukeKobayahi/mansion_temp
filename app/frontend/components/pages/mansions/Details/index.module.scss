@import "~/styles/utils";

.container {
  margin-top: 25px;
  padding: 0 40px;

  @include screenMd {
    padding: 0 16px;
  }
}

.table {
  width: 100%;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  overflow: hidden;

  tr {
    overflow: hidden;
    display: flex;
    width: 100%;
    height: 100%;
    border-bottom: 1px solid #d0d0d0;
    font-size: 16px;

    &:first-child {
      border-top: 1px solid #d0d0d0;
    }

    @include screenMd {
      width: 100%;
      max-width: 100%;
      font-size: 14px;

      &:nth-child(2) {
        border-top: none;
      }
    }
  }

  th,
  td {
    display: inline-block;
    padding: 14px;
    text-align: left;
    line-height: 1.4;
    color: #646464;
  }

  th {
    width: 25%;
    background-color: #faf9f7;
  }

  td {
    width: 75%;
    background-color: #fff;

    @include screenMd {
      padding-right: 0;
    }

    button {
      padding: 0;
      background-color: unset;
      text-decoration: underline;
      color: #646464;
    }
  }
}

.note {
  margin-top: 1.5em;
  font-size: 14px;
  line-height: 1.6;
  text-indent: -1em;
  padding-left: 1em;
  color: #646464;

  @include screenMd {
    font-size: 12px;
  }
}
