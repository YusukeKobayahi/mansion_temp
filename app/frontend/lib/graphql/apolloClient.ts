// 参考
// https://github.com/vercel/next.js/tree/canary/examples/with-apollo

import { useMemo } from "react";
import { ApolloClient, ApolloLink, InMemoryCache } from "@apollo/client";
import { createHttpLink } from "apollo-link-http";
import { from } from "apollo-link";
import { createFragmentArgumentLink } from "apollo-link-fragment-argument";
import { mergeDeep, relayStylePagination } from "@apollo/client/utilities";
import { isSSR } from "~/lib/utils";

/* eslint-disable @typescript-eslint/no-explicit-any -- キャッシュの型を事前に決められないため */
type CacheState = Record<string, any> | null;
/* eslint-enable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/naming-convention -- GraphQLのタイプ名 (このconventionに従わない)を指定したいので */
const createCache = () =>
  new InMemoryCache({
    typePolicies: {
      Mansion: {
        keyFields: ["uniqueCode"],
        fields: {
          salesHistories: relayStylePagination(),
        },
      },
    },
  });
/* eslint-enable @typescript-eslint/naming-convention */

const link = from([
  createFragmentArgumentLink(),
  createHttpLink({
    uri: `${process.env.NEXT_PUBLIC_CORE_API_URL}/consumer/graphql`,
    fetchOptions: {
      mode: "cors",
    },
  }),
]);

function createApolloClient() {
  return new ApolloClient({
    ssrMode: isSSR(),
    // See: https://github.com/apollographql/apollo-link/issues/1258#issuecomment-604031488
    // TODO: @apollo/clientバージョン上げたら直るかも
    link: (link as unknown) as ApolloLink,
    cache: createCache(),
    defaultOptions: {
      watchQuery: {
        // SSR時にuseQueryによってリクエストしてレンダリングする場合、
        // getServerSidePropsからキャッシュが渡っている必要がある。(でないと描画されない)
        // そのため、SSR時は常にcacheから取得されるべきで、実装意図を明確にするためPolicyも設定しておく
        // (エラーになるわけではない)
        fetchPolicy: isSSR() ? "cache-only" : "cache-first",
      },
    },
  });
}

let apolloClient: ReturnType<typeof createApolloClient>;

export function initializeApollo(
  initialState: CacheState = null
): ApolloClient<CacheState> {
  const _apolloClient = apolloClient ?? createApolloClient();

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
    const data = mergeDeep(initialState, existingCache);

    // Restore the cache with the merged data
    _apolloClient.cache.restore(data);
  }
  // For SSG and SSR always create a new Apollo Client
  if (isSSR()) return _apolloClient;
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function useApollo(initialState: CacheState): ApolloClient<CacheState> {
  const store = useMemo(() => initializeApollo(initialState), [initialState]);
  return store;
}
