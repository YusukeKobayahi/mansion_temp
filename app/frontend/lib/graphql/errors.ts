export interface GraphQLError {
  graphQLErrors?: {
    message: string;
    extensions?: {
      code: string;
    };
  }[];
}

export const isNotFoundError = (err: GraphQLError): boolean => {
  const graphQLErrors = err.graphQLErrors;
  if (graphQLErrors == null) return false;

  return graphQLErrors[0]?.extensions?.code === "NOT_FOUND";
};
