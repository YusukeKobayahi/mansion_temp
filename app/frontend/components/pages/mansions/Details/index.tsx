import Link from "next/link";
import { buildLibraryPagePath, buildConstructedDate } from "~/lib/utils";
import isEmpty from "lodash/isEmpty";
import compact from "lodash/compact";
import { useMemo } from "react";
import { useDeepCompareMemo } from "use-deep-compare";
import { event } from "~/lib/gtag";

import styles from "./index.module.scss";
import { MsLibraryMansionsDetailsFragment } from "~/graphql/generated";
import { GoogleAnalyticsEvent } from "~/lib/types";

export interface DetailsProps {
  data: MsLibraryMansionsDetailsFragment;
}

const Details: React.FC<DetailsProps> = ({ data }: DetailsProps) => {
  const constructedDate = buildConstructedDate(data.constructedIn, data.age);
  const floor = useMemo(() => {
    return data.groundFloor == null
      ? "-"
      : data.undergroundFloor !== null
      ? `地下${data.undergroundFloor}階 / ${data.groundFloor}階建て`
      : `${data.groundFloor}階建て`;
  }, [data.groundFloor, data.undergroundFloor]);
  const layouts = useDeepCompareMemo(() => {
    return isEmpty(data.layouts) ? "-" : `${data.layouts.join("、 ")}など`;
  }, [data.layouts]);
  const landUseZones = useDeepCompareMemo(() => {
    return isEmpty(data.landUseZones)
      ? "-"
      : `${data.landUseZones.join("、 ")}`;
  }, [data.landUseZones]);
  const unitAmount = useMemo(
    () => (data.unitAmount !== null ? `${data.unitAmount}戸` : "-"),
    [data.unitAmount]
  );
  const clickArea = (label: GoogleAnalyticsEvent["label"]) => {
    event({
      action: "msl_area_click",
      category: "msl_shosai",
      label: label,
      value: 1,
    });
  };
  const address = useDeepCompareMemo(() => {
    return (
      <span>
        <Link
          href={"/prefectures/[jisCode]"}
          as={buildLibraryPagePath("prefectures", data.prefecture.jisCode)}
          prefetch={false}
          passHref
        >
          <button className={styles.link} onClick={() => clickArea("msl_pref")}>
            {data.prefecture.name}
          </button>
        </Link>
        &nbsp;
        <Link
          href={buildLibraryPagePath("cities", data.city.jisCode)}
          prefetch={false}
          passHref
        >
          <button className={styles.link} onClick={() => clickArea("msl_city")}>
            {data.city.name}
          </button>
        </Link>
        &nbsp;
        <Link
          href={{
            pathname: buildLibraryPagePath("cities", data.city.jisCode),
            query: { tw: data.town.id },
          }}
          prefetch={false}
          passHref
        >
          <button onClick={() => clickArea("msl_town")} className={styles.link}>
            {data.town.name}
          </button>
        </Link>
        &nbsp;
        {compact([data.street?.name, data.blockNumber]).join("")}
      </span>
    );
  }, [
    data.prefecture.name,
    data.city.name,
    data.town.name,
    data.street?.name,
    data.blockNumber,
  ]);
  const accessInfo = useDeepCompareMemo(() => {
    return data.access.map((a, i) => {
      return (
        <span key={i}>
          <Link
            href={buildLibraryPagePath("lines", String(a.lineId))}
            prefetch={false}
            passHref
          >
            <button onClick={() => clickArea("msl_line")}>
              {a.stationLineName}
            </button>
          </Link>
          「
          <Link
            href={buildLibraryPagePath("stations", String(a.stationId))}
            prefetch={false}
          >
            <button onClick={() => clickArea("msl_station")}>
              {a.stationName}
            </button>
          </Link>
          」
          {a.stationWalkingMinutes != null &&
            "徒歩" + a.stationWalkingMinutes + "分"}
        </span>
      );
    });
  }, [data.access]);

  return (
    <section className={styles.container}>
      <table className={styles.table}>
        <tbody className="pcOnly">
          <tr>
            <th>物件名</th>
            <td>{data.name}</td>
            <th>住所</th>
            <td>{address}</td>
          </tr>
          <tr>
            <th>最寄り駅</th>
            <td>{accessInfo}</td>
            <th>築年月</th>
            <td>{constructedDate}</td>
          </tr>
          <tr>
            <th>階建</th>
            <td>{floor}</td>
            <th>間取り</th>
            <td>{layouts}</td>
          </tr>
          <tr>
            <th>総戸数</th>
            <td>{unitAmount}</td>
            <th>構造</th>
            <td>{data.structure.name}</td>
          </tr>
          <tr>
            <th>土地権利</th>
            <td>{data.landPrivilege || "-"}</td>
            <th>用途地域</th>
            <td>{landUseZones}</td>
          </tr>
          <tr>
            <th>施工会社</th>
            <td>{data.constructCompany || "-"}</td>
            <th>管理会社</th>
            <td>{data.managementCompany || "-"}</td>
          </tr>
          <tr>
            <th>管理形態</th>
            <td>{data.managementForm || "-"}</td>
            <th>駐車場</th>
            <td>{data.parking || "-"}</td>
          </tr>
        </tbody>

        <tbody className="spOnly">
          <tr>
            <th>物件名</th>
            <td>{data.name}</td>
          </tr>
          <tr>
            <th>住所</th>
            <td>{address}</td>
          </tr>
          <tr>
            <th>最寄り駅</th>
            <td>{accessInfo}</td>
          </tr>
          <tr>
            <th>階建</th>
            <td>{floor}</td>
          </tr>
          <tr>
            <th>築年月</th>
            <td>{constructedDate}</td>
          </tr>
          <tr>
            <th>間取り</th>
            <td>{layouts}</td>
          </tr>
          <tr>
            <th>総戸数</th>
            <td>{unitAmount}</td>
          </tr>
          <tr>
            <th>構造</th>
            <td>{data.structure.name}</td>
          </tr>
          <tr>
            <th>土地権利</th>
            <td>{data.landPrivilege || "-"}</td>
          </tr>
          <tr>
            <th>用途地域</th>
            <td>{landUseZones}</td>
          </tr>
          <tr>
            <th>施工会社</th>
            <td>{data.constructCompany || "-"}</td>
          </tr>
          <tr>
            <th>管理会社</th>
            <td>{data.managementCompany || "-"}</td>
          </tr>
          <tr>
            <th>管理形態</th>
            <td>{data.managementForm || "-"}</td>
          </tr>
          <tr>
            <th>駐車場</th>
            <td>{data.parking || "-"}</td>
          </tr>
        </tbody>
      </table>
      <p className={styles.note}>
        ※駐車場については過去の売却情報をベースに表記しており、空きがあることを保証するものではありません。
      </p>
    </section>
  );
};

export default Details;
