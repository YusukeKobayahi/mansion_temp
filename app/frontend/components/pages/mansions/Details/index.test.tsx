import { render } from "@testing-library/react";
import mockData from "../fixtures/mockData.json";
import { MansionQuery } from "~/graphql/generated";
import Details from ".";

describe("components/pages/mansions/Details", () => {
  it("正常に表示される", () => {
    const mock = mockData as MansionQuery;
    const { asFragment } = render(<Details data={mock.mansion} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
