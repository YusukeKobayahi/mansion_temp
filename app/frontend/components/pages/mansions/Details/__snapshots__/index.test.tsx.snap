// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/mansions/Details 正常に表示される 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <table
      class="table"
    >
      <tbody
        class="pcOnly"
      >
        <tr>
          <th>
            物件名
          </th>
          <td>
            リヴシティ神田
          </td>
          <th>
            住所
          </th>
          <td>
            <span>
              <button
                class="link"
                href="/prefectures/13000"
              >
                東京都
              </button>
               
              <button
                class="link"
                href="/cities/13101"
              >
                千代田区
              </button>
               
              <button
                class="link"
                href="/cities/13101?tw="
              >
                神田富山町
              </button>
               １６－１
            </span>
          </td>
        </tr>
        <tr>
          <th>
            最寄り駅
          </th>
          <td>
            <span>
              <button
                href="/lines/63"
              >
                山手線
              </button>
              「
              <button>
                神田
              </button>
              」徒歩4分
            </span>
          </td>
          <th>
            築年月
          </th>
          <td>
            2011年03月 (築10年)
          </td>
        </tr>
        <tr>
          <th>
            階建
          </th>
          <td>
            13階建て
          </td>
          <th>
            間取り
          </th>
          <td>
            1DK、 1Kなど
          </td>
        </tr>
        <tr>
          <th>
            総戸数
          </th>
          <td>
            35戸
          </td>
          <th>
            構造
          </th>
          <td>
            鉄筋鉄骨コンクリート造(SRC造)
          </td>
        </tr>
        <tr>
          <th>
            土地権利
          </th>
          <td>
            所有権
          </td>
          <th>
            用途地域
          </th>
          <td>
            商業地域
          </td>
        </tr>
        <tr>
          <th>
            施工会社
          </th>
          <td>
            株式会社谷口工務店
          </td>
          <th>
            管理会社
          </th>
          <td>
            その他
          </td>
        </tr>
        <tr>
          <th>
            管理形態
          </th>
          <td>
            全部委託
          </td>
          <th>
            駐車場
          </th>
          <td>
            無し
          </td>
        </tr>
      </tbody>
      <tbody
        class="spOnly"
      >
        <tr>
          <th>
            物件名
          </th>
          <td>
            リヴシティ神田
          </td>
        </tr>
        <tr>
          <th>
            住所
          </th>
          <td>
            <span>
              <button
                class="link"
                href="/prefectures/13000"
              >
                東京都
              </button>
               
              <button
                class="link"
                href="/cities/13101"
              >
                千代田区
              </button>
               
              <button
                class="link"
                href="/cities/13101?tw="
              >
                神田富山町
              </button>
               １６－１
            </span>
          </td>
        </tr>
        <tr>
          <th>
            最寄り駅
          </th>
          <td>
            <span>
              <button
                href="/lines/63"
              >
                山手線
              </button>
              「
              <button>
                神田
              </button>
              」徒歩4分
            </span>
          </td>
        </tr>
        <tr>
          <th>
            階建
          </th>
          <td>
            13階建て
          </td>
        </tr>
        <tr>
          <th>
            築年月
          </th>
          <td>
            2011年03月 (築10年)
          </td>
        </tr>
        <tr>
          <th>
            間取り
          </th>
          <td>
            1DK、 1Kなど
          </td>
        </tr>
        <tr>
          <th>
            総戸数
          </th>
          <td>
            35戸
          </td>
        </tr>
        <tr>
          <th>
            構造
          </th>
          <td>
            鉄筋鉄骨コンクリート造(SRC造)
          </td>
        </tr>
        <tr>
          <th>
            土地権利
          </th>
          <td>
            所有権
          </td>
        </tr>
        <tr>
          <th>
            用途地域
          </th>
          <td>
            商業地域
          </td>
        </tr>
        <tr>
          <th>
            施工会社
          </th>
          <td>
            株式会社谷口工務店
          </td>
        </tr>
        <tr>
          <th>
            管理会社
          </th>
          <td>
            その他
          </td>
        </tr>
        <tr>
          <th>
            管理形態
          </th>
          <td>
            全部委託
          </td>
        </tr>
        <tr>
          <th>
            駐車場
          </th>
          <td>
            無し
          </td>
        </tr>
      </tbody>
    </table>
    <p
      class="note"
    >
      ※駐車場については過去の売却情報をベースに表記しており、空きがあることを保証するものではありません。
    </p>
  </section>
</DocumentFragment>
`;
